src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -1153,7 +1153,8 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
                     // don't emit if the resulting method would contain initializers with errors
                     if (!hasErrors && (hasBody || includeInitializersInBody))
                     {
-                        Debug.Assert(!methodSymbol.IsImplicitInstanceConstructor || !methodSymbol.ContainingType.IsStructType());
+                        Debug.Assert(!(methodSymbol.IsImplicitInstanceConstructor && methodSymbol.ParameterCount == 0) ||
```

Comment:
>IsImplicitInstanceConstructor [](start = 52, length = 29)

I think the condition `IsImplicitInstanceConstructor` generally implies `methodSymbol.ParameterCount == 0`, so you've weakened the assertion in case there is a mismatch there.

Reply:
Not as far as I can tell: http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Symbols/MethodSymbol.cs,fb6aebe573e152e2,references

That is satisfied for every record constructor as well, so all record constructors are now implicit instance constructors

Reply:
Are there any other implicit instance constructors though? AFAIK, there's only the default parameterless one.

Reply:
Record constructors!

Reply:
Fair enough.

Reply:
Thankfully there's only one use of this property (in this assert), so we don't have to worry about that contract being violated elsewhere.